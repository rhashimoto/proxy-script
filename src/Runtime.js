// Make sure some definitions exist.
// @ts-ignore
const btoa = globalThis.btoa ?? (s => Buffer.from(s).toString('base64'));
const AsyncFunction = (async () => {}).constructor;

// Enumerate global objects and their properties.
const GLOBALS = new WeakSet();
getGlobals(globalThis);

// The AsyncFunction constructor is *not* a global object but
// it needs to be marked as global to prevent mutation.
getGlobals(AsyncFunction);

function getGlobals(obj, key) {
  if (obj === Object(obj) && !GLOBALS.has(obj)) {
    GLOBALS.add(obj);
    const descriptors = Object.getOwnPropertyDescriptors(obj);
    for (const [name, descriptor] of Object.entries(descriptors)) {
      try {
        // Skip prototype getters, which may fail and generally don't
        // return persistent objects anyway.
        if (key !== 'prototype' || !descriptor.get) {
          getGlobals(obj[name], name);
        }
      } catch (e) {
        // @ts-ignore
        if (typeof process === 'undefined') {
          // These messages are suppressed outside the browser environment.
          console.warn('global enumeration failure', name);
        }
      }
    }
  }
}

// Determine number of lines inserted by Function constructor. We need
// this information to patch the sourcemap for browser Dev Tools, and
// it varies with the Javascript implementation.
const SOURCEMAP_OFFSET = (function() {
  const lines = new Function('debugger').toString().split('\n');
  return lines.findIndex(line => line.includes('debugger'));
})();

// Transpiler output is post-processed so its sourcemaps work properly.
// Post-processed output is tagged with this Symbol.
const PREPARED = Symbol('prepared');

export class Runtime {
  /**
   * Whitelist functions can be called. Note that non-whitelist objects
   * (that are not on the blacklist) allow access to properties.
   * @type {Set<function|object>}
   */
  whitelist = new Set([
    [
      Array, Object, Boolean, Error,
      Number, BigInt, Date,
      String, RegExp,
      Map, Set, WeakMap, WeakSet,
      Symbol,
      Promise,

      ArrayBuffer,
      Int8Array, Int16Array, Int32Array,
      Uint8Array, Uint16Array, Uint32Array,
      Float32Array, Float64Array,
      DataView
    ].map(cls => getClassEntries(cls)).flat(),

    // Don't include Function constructor or toString().
    Function.prototype.call,
    Function.prototype.apply,
    Function.prototype.bind,

    getOwnProperties(console),
    getOwnProperties(Math),
    getOwnProperties(JSON),

    globalThis.atob, globalThis.btoa,
    globalThis.isFinite, globalThis.isNaN,
    globalThis.parseFloat, globalThis.parseInt,
    globalThis.encodeURI, globalThis.encodeURIComponent,
    globalThis.decodeURI, globalThis.decodeURIComponent
  ].flat().filter(x => x === Object(x)));

  /**
   * Blacklist objects cannot be used at all.
   * @type {Set<function|object>}
   */
  blacklist = new Set([
    globalThis.document
  ].filter(x => x === Object(x)));

  // All proxies are remembered in this map. This is not just an optimization;
  // it is important that the same proxy is always used for the same object.
  // Consider the case where objects are kept in a set and tested for
  // membership - that doesn't work if there can be multiple proxies.
  //
  // Proxies are mapped to themselves. Hopefully that doesn't affect garbage
  // collection.
  mapObjectToProxy = new WeakMap();

  /** @type {ProxyHandler} */
  handler = {
    apply: (target, thisArg, args) => {
      this.assertAccess(target, 'apply');
      return Reflect.apply(target, thisArg, args);
    },

    construct: (target, args, newTarget) => {
      this.assertAccess(target, 'construct');
      // https://stackoverflow.com/a/49739161/1462337
      return Reflect.construct(target, args, newTarget);
    },

    get: (target, property) => {
      const member = Reflect.get(target, property);
      if (member === Object(member)) {
        // This member object has been reached via a global so it is also
        // marked as a global. This is needed to wrap objects reached via
        // getters, notably DOM objects.
        GLOBALS.add(member);
      }
      return member;
    },

    defineProperty() { throw new Error('proxy-script defineProperty violation'); },
    deleteProperty() { throw new Error('proxy-script deleteProperty violation'); },
    preventExtensions() { throw new Error('proxy-script preventExtensions violation'); },
    set() { throw new Error('proxy-script set violation'); },
    setPrototypeOf() { throw new Error('proxy-script setPrototypeOf violation'); },
  };
  
  /**
   * Runtime Proxy wrapper.
   * @param {object} obj 
   * @returns {any} Proxy for a global, otherwise the argument itself.
   */
  maybeWrap(obj) {
    if (obj !== Object(obj)) return obj;
    if (this.blacklist.has(obj)) throw new Error('proxy-script blacklist violation');
    if (!GLOBALS.has(obj)) return obj;

    if (this.mapObjectToProxy.has(obj)) {
      // This object already has a Proxy so use the existing one.
      return this.mapObjectToProxy.get(obj);
    }

    const proxy = new Proxy(obj, this.handler);
    this.mapObjectToProxy.set(obj, proxy);
    this.mapObjectToProxy.set(proxy, proxy);
    return proxy;
  }
  
  /**
   * Runtime support for method invocation.
   * @param {*} object 
   * @param {*} property 
   * @param  {...any} args 
   */
  call(object, property, ...args) {
    return this.maybeWrap(object[property]).apply(object, args);
  }

  assertAccess(obj, type = 'access') {
    if (GLOBALS.has(obj) && !this.whitelist.has(obj)) {
      throw new Error(`proxy-script ${type} violation`);
    }
  }

  /**
   * @param {{code: string, map?: object}} transpiled 
   */
  prepare(transpiled) {
    if (transpiled[PREPARED]) return transpiled;

    // Add an inline sourcemap if possible.
    let code = transpiled.code;
    const map = transpiled.map && Object.assign({}, transpiled.map);
    if (map && map.version === 3) {
      // Patch sourcemap for insertions by Function constructor.
      // We only need to prepend a semicolon to `mappings` for each line
      // of code added at the start.
      // https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit
      map.mappings = new Array(SOURCEMAP_OFFSET).fill(';').join('') + map.mappings;

      // Add the sourcemap inline to the code.
      const json = JSON.stringify(map);
      const url = `data:application/json;charset=utf-8;base64,${btoa(json)}`;
      if (url.length < 2 ** 23) {
        code += `\n//# sourceMappingURL=${url}`;
      } else {
        console.warn(`proxy-script sourcemap omitted due to size (${url.length} bytes)`);
      }
    }
    return { code, map, [PREPARED]: true };
  }

  /**
   * @param {{code: string, map?: object}} transpiled 
   * @param {object?} [thisArg]
   * @param {object} [args] 
   */
  async runPrepared(transpiled, thisArg, args = {}) {
    console.assert(transpiled[PREPARED], 'running unprepared code');
    this.whitelist.add = () => {
      throw new Error('whitelist modification after run');
    };
    this.blacklist.add(Function);
    this.blacklist.add(AsyncFunction);
    this.blacklist.add(eval);

    args = Object.assign({}, args, {
      '_wrap': this.maybeWrap.bind(this),
      '_call': this.call.bind(this)
    });
    const f = new Function(...Object.keys(args), transpiled.code);
    return await f.call(thisArg, ...Object.values(args));
  }

  /**
   * @param {{code: string, map?: object}} transpiled 
   * @param {object?} thisArg
   * @param {object} args 
   * @returns {Promise}
   */
  run(transpiled, thisArg, args) {
    if (!transpiled[PREPARED]) {
      transpiled = this.prepare(transpiled);
    }
    return this.runPrepared(transpiled, thisArg, args);
  }
};

Runtime.Error = class extends Error {
  constructor(message, { line, column }) {
    super(`${message} (${line}:${column})`);
    this.line = line;
    this.column = column;
  }
};

/**
 * Update collection of global objects in case of additions to the
 * global scope after module load.
 */
Runtime.updateGlobals = () => {
  getGlobals(globalThis);
}

/**
 * Given a class, returns array of related items (including the class).
 * @param {function} cls 
 */
function getClassEntries(cls) {
  return [cls, getOwnProperties(cls), getOwnProperties(cls.prototype)].flat();
}

/**
 * Given an object, returns array of property values.
 * @param {object} obj 
 */
function getOwnProperties(obj) {
  return Object.getOwnPropertyNames(obj).filter(name => {
    const descriptor = Object.getOwnPropertyDescriptor(obj, name);
    return !descriptor.get;
  }).map(name => obj[name]);
}
