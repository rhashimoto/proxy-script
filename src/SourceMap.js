// Determine number of lines inserted by Function constructor. We need
// this information to patch the sourcemap for browser Dev Tools, and
// it varies with the Javascript implementation.
const SOURCEMAP_OFFSET = (function() {
  const lines = new Function('debugger').toString().split('\n');
  return lines.findIndex(line => line.includes('debugger'));
})();

export class SourceMap {
  _groups = [];

  constructor(map) {
    // https://docs.google.com/document/d/1U1RGAehQwRypUTovF1KRlpiOFze0b-_2gc6fAH0KY0k/edit
    /** @type {number?} */ let sourcesIndex = null;
    /** @type {number?} */ let sourceLine = null;
    /** @type {number?} */ let sourceColumn = null;
    /** @type {number?} */ let namesIndex = null;
    map.mappings.split(';').forEach(group => {
      /** @type {number?} */ let startColumn = null;
      this._groups.push(group.split(',').map(segment => {
        const fields = parseSegment(segment);
        switch (fields.length) {
          case 5: namesIndex   += fields[4];
          case 4: sourceColumn += fields[3];
          case 3: sourceLine   += fields[2];
          case 2: sourcesIndex += fields[1];
          case 1: startColumn  += fields[0];
          case 0:
            break;
        }
        return {
          startColumn,
          source: map.sources[sourcesIndex],
          sourceLine,
          sourceColumn,
          name: map.names[namesIndex]
        };
      }));

      // Sort segments in a group by decreasing startColumn for easier lookup.
      this._groups.forEach(group => group.sort((a, b) => {
        return a.startColumn >= b.startColumn ? -1 : 1;
      }));
    });
  }

  /**
   * Returns original source zero-based location.
   * @param {number} line zero-based line in generated code
   * @param {number} column zero-based column in generated code
   */
  locate(line, column) {
    const group = this._groups[line] || [];
    return group.find(segment => segment.startColumn <= column);
  }

  patchStackTrace(trace) {
    return trace.split('\n')
      // TODO: Add processing for Firefox. Edge is Chromium/V8 based so
      // should already work (but should be checked). Safari doesn't have
      // useful stack traces for `new Function()`.
      .map(frame => this.patchV8(frame))
      .join('\n');
  }

  patchV8(frame) {
    const m = frame.match(/^(.*eval at run .*)<anonymous>:(\d+):(\d+)/);
    if (m) {
      // Note adjustments for 1-based line numbers.
      const line = parseInt(m[2]) - 1;
      const column = parseInt(m[3]);
      const location = this.locate(line, column);
      if (location) {
        return `${m[1]}<${location.source}>:${location.sourceLine + 1}:${location.sourceColumn})`;
      }
    }
    return frame;
  }
}

/**
 * Patch sourcemap for insertions by Function constructor.
 * @param {object} map source map
 */
SourceMap.patchMapForFunctionWrapper = function(map) {
  // Prepend a semicolon to `mappings` for each line of code added at
  // the start.
  map.mappings = new Array(SOURCEMAP_OFFSET).fill(';').join('') + map.mappings;
}

/**
 * Build an inline map with a data URL.
 * @param {*} map source map
 */
SourceMap.createInline = function(map) {
  // @ts-ignore
  const btoa = globalThis.btoa ?? (s => Buffer.from(s).toString('base64'));
  const json = JSON.stringify(map);
  const url = `data:application/json;charset=utf-8;base64,${btoa(json)}`;
  return `//# sourceMappingURL=${url}`;
}

/**
 * Translate stack trace locations through a source map.
 * @param {string} trace stack trace from Error.stack
 * @param {object} map source map
 * @returns {string} patched stack strace
 */
SourceMap.patchStackTrace = (trace, map) => {
  const sourceMap = new SourceMap(map);
  return sourceMap.patchStackTrace(trace);
};

/**
 * Map a base64 digit to its numeric value.
 * @type {Map<string, number>}
 */
const BASE_64 = new Map([
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
  'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
  'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
  'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z',
  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/'
].map((c, i) => [c, i]));

/**
 * @param {string} segment 
 * @returns {Array<number>} segment numeric fields
 */
export function parseSegment(segment) {
  const fields = [];

  // Parse VLQs (top bit of digit is continuation bit, LSB of first
  // digit is sign bit). This code only works to 31-bit values instead
  // of the specified 32-bits, which is not expected to be a problem.
  let nBits = 0;
  let value = 0;
  for (const c of segment) {
    const data = BASE_64.get(c);
    if (data === undefined) throw new Error('invalid base64');

    value += (data & 0x1f) << nBits;
    nBits += 5;
    if ((data & 0x20) === 0) {
      const sign = (value & 0x1) ? -1 : 1;
      fields.push(sign * (value >>> 1));
      nBits = value = 0;
    }
  }

  return fields;
}