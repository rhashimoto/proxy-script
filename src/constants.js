export const DEFAULT_GLOBAL_CLASSES = [
  'Array', 'Object', 'Boolean', 'Error',
  'Number', 'BigInt', 'Date',
  'String', 'RegExp',
  'Map', 'Set', 'WeakMap', 'WeakSet',
  'Symbol',
  'Promise',

  'ArrayBuffer',
  'Int8Array', 'Int16Array', 'Int32Array',
  'Uint8Array', 'Uint16Array', 'Uint32Array',
  'Float32Array', 'Float64Array',
  'DataView'
];

export const DEFAULT_GLOBAL_OBJECTS = [
  'console', 'Math', 'JSON'
]

export const DEFAULT_GLOBAL_FUNCTIONS = [
  'atob', 'btoa', 'isFinite', 'isNaN', 'parseFloat', 'parseInt'
].filter(name => globalThis.hasOwnProperty(name));
