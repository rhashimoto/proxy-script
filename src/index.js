import { Transpiler } from './Transpiler.js';
import { Runtime } from './Runtime.js';

export { Transpiler, Runtime };
