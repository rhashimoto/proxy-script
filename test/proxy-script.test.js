// Copyright 2021 Roy T. Hashimoto. All rights reserved.

import * as Babel from '@babel/standalone';
import { Transpiler, Runtime } from '../src/index';

globalThis.Babel = Babel;
Transpiler.register(Babel);

describe('proxy-test', () => {
  test('transpiler works with no options', () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('1 + 1');
    expect(transpiled.code).toBeDefined();
  });

  test('transpiler uses fixed wrapper alias', () => {
    const transpiler = new Transpiler({
      bindings: {
        wrap: 'myWrap',
        call: 'myCall',
        func: 'myFunc',
        klass: 'myKlass',
        external: 'myExternal'
      }
    });
    const transpiled = transpiler.transpile('({}).foo');
    expect(transpiled.code).toMatch(/myWrap/);
    expect(transpiled.code).toMatch(/myCall/);
    expect(transpiled.code).toMatch(/myFunc/);
    expect(transpiled.code).toMatch(/myKlass/);
    expect(transpiled.code).toMatch(/myExternal/);
    expect(transpiled.code).toMatchSnapshot();
  });

  test('runtime returns value', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return 42;');

    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe(42);
  });

  test('runtime passes arguments', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return a + b;');

    const runtime = new Runtime();
    const result = runtime.run(transpiled, undefined, { a: 1, b: 2 });
    await expect(result).resolves.toBe(3);
  });

  test('runtime passes target (thisArg)', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return this.foo;');

    const runtime = new Runtime();
    const result = runtime.run(transpiled, { foo: 42 });
    await expect(result).resolves.toBe(42);
  });

  test('runtime allows whitelisted free function call', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return isFinite(3.14);');

    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe(true);
  });

  test('runtime rejects non-whitelisted free function call', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return isFinite(3.14);');

    const runtime = new Runtime();
    runtime.whitelist.delete(isFinite);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime allows whitelisted static object call', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return Math.pow(2, 8);');

    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe(256);
  });

  test('runtime rejects non-whitelisted static object call', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return Math.pow(2, 8);');

    const runtime = new Runtime();
    runtime.whitelist.delete(Math.pow);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime allows whitelisted construction', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return new Array();');

    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBeDefined();
  });

  test('runtime rejects non-whitelisted construction', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return new Array();');

    const runtime = new Runtime();
    runtime.whitelist.delete(Array);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime allows whitelisted methods', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      const m = new Map();
      m.set('foo', 'bar');
      return m.get('foo');
    `);

    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe('bar');
  });

  test('runtime rejects non-whitelisted methods', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      const m = new Map();
      m.set('foo', 'bar');
      return m.get('foo');
    `);

    const runtime = new Runtime();
    runtime.whitelist.delete(Map.prototype.set);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
});

  test('runtime rejects global object mutation', async () => {
    {
      // Assigning to an existing property.
      const transpiler = new Transpiler();
      const transpiled = transpiler.transpile('Object.assign = null;');
      const runtime = new Runtime();
      const result = runtime.run(transpiled);
      await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
    }  
    {
      // Assigning to a new property.
      const transpiler = new Transpiler();
      const transpiled = transpiler.transpile('Object.brandNewProperty = null;');
      const runtime = new Runtime();
      const result = runtime.run(transpiled);
      await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
    }  
  });

  test('runtime allows whitelisted indirect call', async () => {
    // Attempt to access the Object constructor via a literal.
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return ({}).constructor()');

    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toEqual({});
  });

  test('runtime rejects non-whitelisted indirect call', async () => {
    // Attempt to access the Object constructor via a literal.
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile('return ({}).constructor()');

    const runtime = new Runtime();
    runtime.whitelist.delete(Object); // remove Object from whitelist
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime rejects Object.assign() to global', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`Object.assign(Object, { foo: 0 })`);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime rejects calling eval', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      eval('globalThis');
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime rejects Function constructor', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      new Function('globalThis');
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime rejects indirect Function construction', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      new (() => {}).constructor('globalThis')
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime rejects indirect AsyncFunction construction', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      new (async () => {}).constructor('globalThis')
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('shadowing wrapper function fails', async () => {
    // The transpiler adds a wrapper function around every expression
    // that could be a global object. If we can redefine the wrapper
    // function then the Proxy guards are bypassed.
    //
    // This doesn't work because the wrapper function has a randomly
    // generated name every time code is transpiled.
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      (function () {
        // Attempt to shadow the wrapper function with a no-op.
        const _wrapWithProxy = x => x;
        Object.foo = 'foo';
      })();
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('function arguments are access checked', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      function foo(a) {
        return a(2, 4);
      }
      return foo(Math.pow);
    `);
    const runtime = new Runtime();
    runtime.whitelist.delete(Math.pow);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('code inside functions is access checked', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      function foo() {
        return Math.pow(2, 4);
      }
      return foo();
    `);
    const runtime = new Runtime();
    runtime.whitelist.delete(Math.pow);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('Promise inside function works', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      async function foo() {
        return await Promise.resolve(42);
      }
      return foo();
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe(42);
  });

  test('Function.prototype.call method works for whitelisted function', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      return Array.prototype.map.call('abc', x => x);
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toEqual(['a', 'b', 'c']);
  });

  test('Function.prototype.apply method works for whitelisted function', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      return Array.prototype.map.apply('abc', [x => x]);
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toEqual(['a', 'b', 'c']);
  });

  test('Function.prototype.call method rejects for non-whitelisted function', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      return Array.prototype.map.call('abc', x => x);
    `);
    const runtime = new Runtime();
    runtime.whitelist.delete(Array.prototype.map);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('Function.prototype.apply method rejects for non-whitelisted function', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      return Array.prototype.map.apply('abc', [x => x]);
    `);
    const runtime = new Runtime();
    runtime.whitelist.delete(Array.prototype.map);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('runtime allows nested Promise.resolve() inside function', async () => {
    // The issue here is that V8 (at least) can be temperamental about
    // accepting a Proxy of a Promise as a Promise.
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      function foo() {
        return Promise.resolve(Promise.resolve(42));
      }
      return foo();
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe(42);
  });

  test('runtime rejects Function subclass', async () => {
    // If we can manage to create a function from a string then that
    // provides unrestricted access to the global scope. A direct call
    // to `new Function(...)` isn't allowed, but how about subclassing?
    //
    // This is prevented because here we actually subclass a Proxy, so
    // the call via `super()` is intercepted.
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      class Foo extends Function {
        constructor(source) {
          super(source);
        }
      };

      const foo = new Foo('return 42;');
      return foo();
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('prefix/postfix update (++ and --)', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      let i = 0;
      let j = ++i;
      let k = j--;
      return [i, j, k];
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toEqual([1, 0, 1]);
  });

  test('using with to shadow wrapper fails', async () => {
    // The idea here is to extract the random wrapper alias by examining
    // the code's own source, then setting up a `with` block that shadows
    // the wrapper.
    //
    // This fails for several reasons. Both the transpiler and runtime
    // use strict mode, so `arguments.callee` and `with` are not allowed.
    // Also, the default whitelist and settings don't include
    // `Function.prototype.toString`.
    const transpiler = new Transpiler({ bindings: { wrap: '$$' } });
    transpiler.babelOptions.parserOpts.strictMode = false;
    const transpiled = transpiler.transpile(`
      const text = arguments.callee.toString();
      const m = text.match(/(\w+) = _wrapWithProxy/);
      const wrapperName = m[1];
      with ({ [wrapperName]: x => x }) {
        // evil code goes here
      }
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).rejects.toThrowErrorMatchingSnapshot();
  });

  test('literal member expression', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      return 'abc'.repeat(3);
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe('abcabcabc');
  });

  test('Object.defineProperty() works on user object', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      class MyMap extends Map {
        constructor() { super(); }
      }
      Object.defineProperty(MyMap.prototype, 'foo', {
        value: 42
      });

      const obj = new MyMap();
      return obj.foo;
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe(42);
  });

  test('Object.defineProperty() rejects on global object', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      const MyMap = Map;
      Object.defineProperty(MyMap.prototype, 'foo', {
        value: 42
      });

      const obj = new MyMap();
      return obj.foo;
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('Object.deleteProperty() works on user object', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      class MyMap {
        has() { return true; }
      }

      const obj = new MyMap();
      delete MyMap.prototype.has;

      return 'has' in obj;
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toBe(false);
  });

  test('Object.deleteProperty() rejects on global object', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      const MyMap = Map;

      const obj = new MyMap();
      delete MyMap.prototype.has;

      return 'has' in obj;
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('map succeeds with whitelisted functions', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      return [42, '42', Math.sqrt(-1)].map(isNaN);
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toEqual([false, false, true]);
  });

  test('map rejects with non-whitelisted functions', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      return [42, '42', Math.sqrt(-1)].map(isNaN);
    `);
    const runtime = new Runtime();
    runtime.whitelist.delete(isNaN);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('bind succeeds with whitelisted functions', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      const f = Array.prototype.reverse.bind([1, 2, 3]);
      return f();
    `);
    const runtime = new Runtime();
    const result = runtime.run(transpiled);
    await expect(result).resolves.toEqual([3, 2, 1]);
  });

  test('bind rejects with non-whitelisted functions', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      const f = Array.prototype.reverse.bind([1, 2, 3]);
      return f();
    `);
    const runtime = new Runtime();
    runtime.whitelist.delete(Array.prototype.reverse);
    const result = runtime.run(transpiled);
    await expect(result.catch(e => e)).resolves.toBeInstanceOf(Runtime.Error);
  });

  test('Error locations are translated with source map', async () => {
    const transpiler = new Transpiler();
    const transpiled = transpiler.transpile(`
      // This comment is line 2.
      throw new Error('this is line 3');
    `);

    const runtime = new Runtime();
    const result = await runtime.run(transpiled).catch(e => e);
    expect(result.stack).toMatch(/<proxy-script>:3/);
  });
});
