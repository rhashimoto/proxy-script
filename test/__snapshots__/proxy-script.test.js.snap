// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proxy-test transpiler uses fixed wrapper alias 1`] = `
"\\"use strict\\";

const myWrap = _wrap;
const myCall = _call;
const myFunc = _func;
const myKlass = _klass;
const myExternal = _external;
return (async () => {
  myWrap({}).foo;
})();"
`;
